### PR TITLE
*: Speed up the add index operation

### PR DESCRIPTION
**Pull request by zimulala**
Using 4 TiKV, 1 PD, 2 TiDB , the servers of TiKV and TiDB are not on a computer.
The number of records in this table is 100,000,000, and the table's structure is as follows:
 CREATE TABLE `tt` ( 
 `a` int(11) DEFAULT NULL,
  `b` varchar(256) DEFAULT NULL,
  `c` varchar(256) DEFAULT NULL,
  `d` int(11) DEFAULT NULL,
  `e` int(11) DEFAULT NULL,
  `f` int(11) DEFAULT NULL)ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin

tidb-before: adding an index e takes time 44m33.015285749s
tidb-current: adding an index f takes time 35m33.008533773s

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @coocood @winkyao @breeswish @tiancaiamao

File: ddl/index.go [hunk 1]
```diff
@@ -532,138 +549,101 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
...
+	baseHandle := reorgInfo.Handle
 
+	workers := make([]*worker, workerCnt)
+	for i := 0; i < workerCnt; i++ {
+		ctx := d.newContext()
+		workers[i] = newWorker(ctx, i, int(taskBatch), len(cols), len(taskOpInfo.colMap))
+	}
 	for {
 		startTime := time.Now()
+		doneHandle := baseHandle
```
Comment (coocood) [inline ddl/index.go]: This line can be removed?

**Comment by zimulala** [comment]
PTAL @coocood

File: ddl/index.go [hunk 2]
```diff
@@ -532,138 +549,100 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
...
-	l := len(taskOpInfo.taskRetCh)
-	taskRets := make([]*taskResult, 0, l)
-	for i := 0; i < l; i++ {
-		taskRet := <-taskOpInfo.taskRetCh
-		taskRets = append(taskRets, taskRet)
-	}
-	sort.Sort(taskRetSlice(taskRets))
-
-	taskAddedCount, currHandle := int64(0), int64(0)
+func getCountAndHandle(prevDoneHandle int64, workers []*worker) (int64, int64, bool, error) {
```
Comment (coocood) [inline ddl/index.go]: The variable `prevDoneHandle` can be removed. 
If the first worker failed, currHandle would be 0, then we don't need to update reorgInfo.

File: ddl/index.go [hunk 3]
```diff
@@ -532,138 +549,100 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
...
+				addedCount, taskAddedCount, sub, err1)
 			return errors.Trace(err)
 		}
 		d.setReorgRowCount(addedCount)
 		batchHandleDataHistogram.WithLabelValues(batchAddIdx).Observe(sub)
 		log.Infof("[ddl] total added index for %d rows, this task added index for %d rows, take time %v",
 			addedCount, taskAddedCount, sub)
 
-		if retCnt < taskCnt {
+		prevDoneHandle, baseHandle = doneHandle, doneHandle+1
```
Comment (coocood) [inline ddl/index.go]: We don't need to set `baseHandle` to `doneHandle+1`, it is already the value we need.

File: tablecodec/tablecodec.go [hunk 4]
```diff
@@ -227,6 +227,60 @@ func DecodeColumnValue(data []byte, ft *types.FieldType, loc *time.Location) (ty
 	return colDatum, nil
 }
 
+// DecodeRowWithMap decodes a byte slice into datums with a existing row map.
+// Row layout: colID1, value1, colID2, value2, .....
+func DecodeRowWithMap(b []byte, cols map[int64]*types.FieldType, loc *time.Location, row map[int64]types.Datum) error {
```
Comment (coocood) [inline tablecodec/tablecodec.go]: There are duplicated code in `DecodeRow`, we can make `DecodeRow` call `DecodeRoWIthMap`.

Comment (zimulala) [reply tablecodec/tablecodec.go]: Got it.

File: ddl/index.go (hunk 2 above)
Comment (zimulala) [reply ddl/index.go]: We don't know when we need to update the reorgInfo's handle.

File: ddl/index.go [hunk 5]
```diff
@@ -386,111 +396,90 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
 	defaultSmallBatchCnt = 128
 	defaultTaskHandleCnt = 128
-	defaultTaskCnt       = 16
+	defaultWorkers       = 16
 )
 
 // taskResult is the result of the task.
 type taskResult struct {
 	count      int   // The number of records that has been processed in the task.
 	doneHandle int64 // This is the last reorg handle that has been processed.
```
Comment (coocood) [inline ddl/index.go]: Can we remove the `doneHandle`?
We can use worker's taskRange to update the reorg info handle.

File: ddl/index.go (hunk 2 above)
Comment (coocood) [reply ddl/index.go]: Or we can just use the first worker's `startHandle`

File: ddl/index.go (hunk 3 above)
Comment (zimulala) [reply ddl/index.go]: These handles may be deleted many in bulk,  so the `doneHandle + 1` may be much larger than `baseHandle`.

File: ddl/index.go (hunk 5 above)
Comment (zimulala) [reply ddl/index.go]: `doneHandle` may be much larger than `endHandle`.

File: ddl/index.go [hunk 6]
```diff
@@ -386,111 +396,90 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+		idxVal[j] = idxColumnVal
 	}
+	idxRecord.vals = idxVal
 	return nil
 }
 
 const (
 	defaultBatchCnt      = 1024
 	defaultSmallBatchCnt = 128
 	defaultTaskHandleCnt = 128
```
Comment (coocood) [inline ddl/index.go]: We use Seek to fetch handles and Seek is implement by Scan in TiKV, which has 256 batch size, change it to 256 may improve performance.

File: ddl/index.go (hunk 5 above)
Comment (coocood) [reply ddl/index.go]: OK, so can we change the name to reflect it.
How about `outOfRangeHandle`, then we don't need to `+1`, we can use it directly for the next baseHandle.

File: ddl/index.go (hunk 2 above)
Comment (zimulala) [reply ddl/index.go]: We need to know when seeking is end.

File: ddl/ddl_db_test.go [hunk 7]
```diff
@@ -383,16 +385,18 @@ LOOP:
 			for i := num; i < num+step; i++ {
 				n := rand.Intn(num)
 				deletedKeys[n] = struct{}{}
-				s.mustExec(c, "delete from t1 where c1 = ?", n)
-				s.mustExec(c, "insert into t1 values (?, ?, ?)", i, i, i)
+				sql := fmt.Sprintf("delete from t1 where c1 = %d", n)
```
Comment (coocood) [inline ddl/ddl_db_test.go]: Why change the test?

File: tablecodec/tablecodec.go [hunk 8]
```diff
@@ -227,16 +227,15 @@ func DecodeColumnValue(data []byte, ft *types.FieldType, loc *time.Location) (ty
 	return colDatum, nil
 }
 
-// DecodeRow decodes a byte slice into datums.
+// DecodeRowWithMap decodes a byte slice into datums with a existing row map.
 // Row layout: colID1, value1, colID2, value2, .....
-func DecodeRow(b []byte, cols map[int64]*types.FieldType, loc *time.Location) (map[int64]types.Datum, error) {
-	if b == nil {
+func DecodeRowWithMap(b []byte, cols map[int64]*types.FieldType, loc *time.Location, row map[int64]types.Datum) (map[int64]types.Datum, error) {
+	if row == nil || b == nil {
```
Comment (coocood) [inline tablecodec/tablecodec.go]: How about creating the map if `row` is nil at line 239?
Then we don't need to create the map in `DecodeRow`.

Or we just remove `DecodeRow`, use `DecodeRowWithMap` instead?
Or adding an argument `row` to `DecodeRow`?

File: ddl/ddl_db_test.go (hunk 7 above)
Comment (zimulala) [reply ddl/ddl_db_test.go]: If the test is failed, we can have more detail logs.

**Comment by coocood** [comment]
LGTM

**Pull request merged**